File: R/ap_v3.R
# This function implements an optimized binomial tree algorithm for
# valuing American put options using Cox-Ross-Rubinstein parameters.
# In relation to ap_v2 includes defensive programming with input validation.

ap_v3 <- function(S0, K, r, q, tt, sigma, steps) {
  # S0: Initial stock price
  # K: Strike price
  # r: Risk-free interest rate compounded continuously
  # q: Dividend yield compounded continuously
  # tt: Time to maturity in years
  # sigma: Volatility of the underlying stock
  # steps: Number of time steps in the binomial tree

  # Input validation
  # Check for missing arguments
  if (
    missing(S0) ||
      missing(K) ||
      missing(r) ||
      missing(q) ||
      missing(tt) ||
      missing(sigma) ||
      missing(steps)
  ) {
    stop("All arguments must be provided")
  }

  # Check for NA or NULL values
  if (
    any(is.na(c(S0, K, r, q, tt, sigma, steps))) ||
      any(is.null(c(S0, K, r, q, tt, sigma, steps)))
  ) {
    stop("Arguments cannot be NA or NULL")
  }

  # Check that all inputs are numeric
  if (
    !is.numeric(S0) ||
      !is.numeric(K) ||
      !is.numeric(r) ||
      !is.numeric(q) ||
      !is.numeric(tt) ||
      !is.numeric(sigma) ||
      !is.numeric(steps)
  ) {
    stop("All arguments must be numeric")
  }

  # Check that inputs are scalar (length 1)
  if (
    length(S0) != 1 ||
      length(K) != 1 ||
      length(r) != 1 ||
      length(q) != 1 ||
      length(tt) != 1 ||
      length(sigma) != 1 ||
      length(steps) != 1
  ) {
    stop("All arguments must be scalar values")
  }

  # Validate parameter ranges
  if (S0 <= 0) {
    stop("Initial stock price (S0) must be positive")
  }

  if (K <= 0) {
    stop("Strike price (K) must be positive")
  }

  if (tt <= 0) {
    stop("Time to maturity (tt) must be positive")
  }

  if (sigma <= 0) {
    stop("Volatility (sigma) must be positive")
  }

  if (q < 0) {
    stop("Dividend yield (q) must be non-negative")
  }

  if (steps < 1) {
    stop("Number of steps must be at least 1")
  }

  # Check that steps is an integer
  if (steps != floor(steps)) {
    stop("Number of steps must be an integer")
  }

  # Calculate parameters
  dt <- tt / steps # time between steps
  u <- exp(sigma * sqrt(dt)) # Up factor
  d <- exp(-sigma * sqrt(dt)) # Down factor
  p <- (exp((r - q) * dt) - d) / (u - d) # Risk-neutral probability

  # Validate risk-neutral probability (arbitrage-free condition)
  if (p < 0 || p > 1) {
    stop(sprintf(
      "Risk-neutral probability p=%.4f is outside [0,1]. Model parameters lead to arbitrage.",
      p
    ))
  }

  disc <- exp(-r * dt) # Discount factor

  # Vector to store option values
  option_values <- numeric(steps + 1)

  # Backward induction from maturity to root
  for (j in (steps + 1):1) {
    num_nodes <- if (j == steps + 1) steps + 1 else j

    for (i in 1:num_nodes) {
      # Calculate asset price at this node
      asset_price <- S0 * u^(j - i) * d^(i - 1)

      if (j == steps + 1) {
        # At maturity: option value is intrinsic value
        option_values[i] <- max(K - asset_price, 0)
      } else {
        # Before maturity: calculate continuation value and check early exercise
        continuation_value <- (p *
          option_values[i] +
          (1 - p) * option_values[i + 1]) *
          disc
        intrinsic_value <- K - asset_price
        option_values[i] <- max(continuation_value, intrinsic_value)
      }
    }
  }

  return(option_values[1]) # Option price at the root node
}

File: R/ap_v3.jl
# American Put Option Pricing using Binomial Tree
# Julia implementation optimized for performance

"""
    american_put_binomial(S0, K, r, q, tt, sigma, steps)

Calculate the price of an American put option using the Cox-Ross-Rubinstein binomial tree model.

# Arguments
- `S0::Real`: Initial stock price
- `K::Real`: Strike price
- `r::Real`: Risk-free interest rate (continuously compounded)
- `q::Real`: Dividend yield (continuously compounded)
- `tt::Real`: Time to maturity in years
- `sigma::Real`: Volatility of the underlying stock
- `steps::Integer`: Number of time steps in the binomial tree

# Returns
- `Float64`: Option price at the root node
"""
function american_put_binomial(S0::Real, K::Real, r::Real, q::Real,
                                tt::Real, sigma::Real, steps::Integer)

    # Input validation
    S0 > 0 || throw(ArgumentError("Initial stock price (S0) must be positive"))
    K > 0 || throw(ArgumentError("Strike price (K) must be positive"))
    tt > 0 || throw(ArgumentError("Time to maturity (tt) must be positive"))
    sigma > 0 || throw(ArgumentError("Volatility (sigma) must be positive"))
    q >= 0 || throw(ArgumentError("Dividend yield (q) must be non-negative"))
    steps >= 1 || throw(ArgumentError("Number of steps must be at least 1"))

    # Calculate parameters
    dt = tt / steps  # time between steps
    u = exp(sigma * sqrt(dt))  # Up factor
    d = exp(-sigma * sqrt(dt))  # Down factor
    p = (exp((r - q) * dt) - d) / (u - d)  # Risk-neutral probability

    # Validate risk-neutral probability (arbitrage-free condition)
    if p < 0 || p > 1
        throw(ArgumentError(
            "Risk-neutral probability p=$p is outside [0,1]. Model parameters lead to arbitrage."
        ))
    end

    disc = exp(-r * dt)  # Discount factor

    # Pre-allocate vector to store option values
    option_values = Vector{Float64}(undef, steps + 1)

    # Backward induction from maturity to root
    for j in (steps + 1):-1:1
        num_nodes = (j == steps + 1) ? steps + 1 : j

        for i in 1:num_nodes
            # Calculate asset price at this node
            asset_price = S0 * u^(j - i) * d^(i - 1)

            if j == steps + 1
                # At maturity: option value is intrinsic value
                option_values[i] = max(K - asset_price, 0.0)
            else
                # Before maturity: calculate continuation value and check early exercise
                continuation_value = (p * option_values[i] +
                                     (1 - p) * option_values[i + 1]) * disc
                intrinsic_value = K - asset_price
                option_values[i] = max(continuation_value, intrinsic_value)
            end
        end
    end

    return option_values[1]  # Option price at the root node
end

"""
    american_put_binomial_batch(S0_vec, K_vec, r_vec, q_vec, tt_vec, sigma_vec, steps_vec)

Calculate prices for multiple American put options using multi-threading.
This is much more efficient than calling american_put_binomial in a loop.

# Arguments
All arguments should be vectors of the same length:
- `S0_vec`: Vector of initial stock prices
- `K_vec`: Vector of strike prices
- `r_vec`: Vector of risk-free interest rates
- `q_vec`: Vector of dividend yields
- `tt_vec`: Vector of times to maturity
- `sigma_vec`: Vector of volatilities
- `steps_vec`: Vector of numbers of time steps

# Returns
- `Vector{Float64}`: Vector of option prices
"""
function american_put_binomial_batch(S0_vec::AbstractVector,
                                     K_vec::AbstractVector,
                                     r_vec::AbstractVector,
                                     q_vec::AbstractVector,
                                     tt_vec::AbstractVector,
                                     sigma_vec::AbstractVector,
                                     steps_vec::AbstractVector{<:Integer})
    # Validate that all vectors have the same length
    n = length(S0_vec)
    if length(K_vec) != n || length(r_vec) != n || length(q_vec) != n ||
       length(tt_vec) != n || length(sigma_vec) != n || length(steps_vec) != n
        throw(ArgumentError("All input vectors must have the same length"))
    end

    # Pre-allocate result vector
    results = Vector{Float64}(undef, n)

    # Use multi-threading to process options in parallel
    Threads.@threads for i in 1:n
        results[i] = american_put_binomial(
            S0_vec[i],
            K_vec[i],
            r_vec[i],
            q_vec[i],
            tt_vec[i],
            sigma_vec[i],
            steps_vec[i]
        )
    end

    return results
end
